src/flag_gems/ops/sort.py

```diff
@@ -12,6 +12,306 @@
+            acc = tl.zeros((TILE_R, TILE_N), dtype=tl.int64)
+            for n_start in range(cta_n_start, cta_n_end, TILE_N):  # sequantial
+                n_offsets = n_start + tl.arange(0, TILE_N)  # (TILE_N, )
+                mask = n_offsets < cta_n_end
+                arr = tl.load(arr_ptr + pid_m * n + n_offsets, mask=mask)
+                arr = convert_to_uint_preverse_order(arr, descending)
+                key = (arr >> bit_offset) & bfe_mask  # (TILE_N, )
+                matches = tl.where(
+                    mask, (bin_indices[:, None] == key), False
+                )  # (TILE_R, TILE_N)
+                acc += matches
+            local_sum = tl.sum(acc, axis=1)
+            tl.atomic_add(
+                out_ptr + pid_m * num_passes * r + p * r + bin_indices,
+                local_sum,
+                sem="relaxed",
+            )
+
+
+@triton.jit
+def sweep(
+    arr_ptr,
+    associate_arr_ptr,  # inputs: (key & value)
+    out_ptr,
+    associate_out_ptr,  # outputs: (key & value)
+    excumsum_bins_ptr,
+    status_ptr,  # aux input and status
+    n_passes,
+    pass_id,
+    bit_offset,
+    N,
+    OUT_N,
+    TILE_N: tl.constexpr,
+    TILE_R: tl.constexpr,
+    k_bits: tl.constexpr,
+    descending: tl.constexpr,
+):
+    # r: num_bins = 2 ** k_bits
+    # OUT_N: grid_n = cdiv(N, )
+
+    # arr_ptr: (m, N)
+    # out_ptr: (m, N)
+    # excumsum_bins_ptr: (m, n_passes, r)
+    # flag_ptr: (m, r, OUT_N)
+
+    # grid: (m, grid_r, grid_n)
+
+    # load data
+    pid_n = tl.program_id(1)  # TODO: use a global counter to avoid dead-lock
+    pid_r = tl.program_id(2)
+    pid_m = tl.program_id(0)
+
+    # bit masks
+    aggregate_mask: tl.constexpr = 1 << 30
+    inclusive_prefix_mask: tl.constexpr = 1 << 31
+    v_mask: tl.constexpr = (1 << 30) - 1
+    bfe_mask: tl.constexpr = (1 << k_bits) - 1  # a.k.a. 2 ** k_bits - 1
+
+    # initialize flag to zero-local sum is not ready
+    r: tl.constexpr = 2**k_bits
+    cta_r_start = pid_r * TILE_R
+    cta_r_end = tl.minimum(cta_r_start + TILE_R, r)
+
+    # cumsum for a bin_index
+    n_offsets = pid_n * TILE_N + tl.arange(0, TILE_N)  # (TILE_N, )
+    mask = n_offsets < N
+    arr = tl.load(arr_ptr + pid_m * N + n_offsets, mask=mask)
+    arr_u = convert_to_uint_preverse_order(arr, descending)
+    key = (arr_u >> bit_offset) & bfe_mask  # (TILE_N, )
+
+    # since triton can only use scalar as condition, loop by bin_index
+    # status must be pre zero-initialized, or else we have to initialize it
+    for bin_index in range(cta_r_start, cta_r_end):
+        matches = tl.where(mask, key == bin_index, False)  # (TILE_N, ) bool
+        # cta level cumsum per bin
+        # CAUTION: tl.sum in triton 3.2 does not promote type
+        local_sum = tl.sum(matches.to(tl.uint32), axis=0)
+        pack0 = aggregate_mask | local_sum
+        status_offset = pid_m * (r * OUT_N) + bin_index * OUT_N + pid_n
+        tl.store(status_ptr + status_offset, pack0, cache_modifier=".cg")
```

Comment:
Is it possible to support values of n up to 2^31 by storing the local_sum and a flag bit separately?